### DISCUSSION TITLE
Errors in middlewares - must `api` be global?

### DISCUSSION THREAD

**Discussion by davidolrik** [danielgtaylor/huma#407] [Q&A] [answered]
The docs states that in order to generate an error in a middleware, I have to know the `api` inside the middleware func like so:

```go
func MyMiddleware(ctx huma.Context, next func(ctx huma.Context)) {
    if ctx.Query("error") == "true" {
        huma.WriteErr(
            api,
            ctx,
            http.StatusInternalServerError,
            "Some friendly message", fmt.Errorf("error detail"),
        )
        return
    }

    next(ctx)
}
```

How can I access `api` without making it a global variable?

It seems to me a better approach would be to make huma middlewares return an optional error instead, and then let huma write out the error like it does for request errors, like so:

```go
func MyMiddleware(ctx huma.Context, next func(ctx huma.Context)) *huma.StatusError {
    if ctx.Query("error") == "true" {
        return huma.Error500InternalServerError("Some friendly message")
    }
    return next(ctx)
}
```

Or am I missing something? - Could app be part of the huma context?

**Comment by x-user** [comment, marked as answer]
Just use closures bro.
```go
// ...

func main() {
	// ...
	api.UseMiddleware(MyMiddleware(api))
	// ...
}

// ...

func MyMiddleware(api huma.API) func(ctx huma.Context, next func(huma.Context)) {
	return func(ctx huma.Context, next func(huma.Context)) {
		// ...
		huma.WriteErr(api, ctx, http.StatusNotImplemented, http.StatusText(http.StatusNotImplemented))
	}
}
```

**Comment by NaviteLogger** [comment]
Don't do that. huma reference should not be passed to middleware. This causes issues when multiple async requests hit your middleware

**Reply by gfortia** [reply]
Why? Having a global variable is what is suggested in the docs. Why does creating a global variable not cause the problems you mention but the suggested closure approach does?
